### DISCUSSION TITLE
Why React, not Vue?

### DISCUSSION THREAD

**Discussion by kinnalover** [fastapi/full-stack-fastapi-template#1108] [Questions] [answered]
In previous versions we had Vue as a frontend. Why you switched to React?

**Comment by Haiz14** [comment]
React is the most popular. Quite appropriate choice isn't it? 

If you are a maintainer what will you choose? Something that more users will be familiar with? Or something that less users will be familiar with?

**Reply by Justice-Udehegbunam** [reply]
pls i made an update specifically for this issue, please has it been reviewed

**Comment by khushinj** [comment]
> In previous versions we had Vue as a frontend. Why you switched to React?


Project Requirements: The decision to switch frontend frameworks may be driven by changes in project requirements. React might offer specific features, performance improvements, or integrations that better align with the goals of the project.

    Community Support: React has a large and active community with extensive resources, libraries, and tooling available. If the project team anticipates needing robust community support or integration with other libraries, React could be a more favorable choice.

    Developer Familiarity: Sometimes, the decision to switch frameworks is influenced by the development team's expertise and familiarity. If the team is more experienced or comfortable with React, they might opt to use it for consistency and efficiency in development.

    Ecosystem Considerations: React's ecosystem continues to evolve with new tools and libraries emerging regularly. The availability of specific tools or integrations within the React ecosystem might be a factor in the decision-making process.

    Performance and Scalability: React is known for its performance optimizations and scalability. If the project anticipates significant growth or performance requirements, React might be chosen for its ability to handle complex applications efficiently.

**Comment by judidoko** [comment]
Switching from Vue.js to React often involves a variety of factors and considerations, including technical, strategic, and development Team preferences.
This could be;

- Ecosystem or Library Support: React has a larger ecosystem and community support compared to Vue.js. The project requires specific libraries and integrations that are more readily available in the React ecosystem, so it might make sense to switch.

- Performance Requirements: React's virtual DOM implementation may offer better performance in certain scenarios compared to Vue.js, especially in large-scale applications. If performance becomes a critical factor for the project, the team might opt for React.

- Learning Curve and Developer Pool: React has a steeper learning curve compared to Vue.js, but it's also more widely known among developers. As an open-source project that needs to onboard new developers quickly for more contribution, switching to React might be preferred.

- Integration with Other Technologies: React is often preferred for projects that require seamless integration with other technologies or frameworks, such as GraphQL, TypeScript etc. If the project's requirements evolve to prioritize these integrations, switching to React could be a strategic decision.

- Community and Industry Trends: Sometimes, the decision to switch frameworks is influenced by broader industry trends or community preferences. React is more dominant and there's a shift in developer preferences towards React.

- Scalability and Flexibility: React's component-based architecture and unidirectional data flow make it well-suited for building scalable and flexible applications. The project's scalability requirements have increased over time as the project grows, which leads to switching to React.

**Comment by gauravxdev** [comment]
the transition from Vue.js to React is likely driven by React's stronger capabilities in handling complex applications. React offers a robust ecosystem, modular architecture, and efficient rendering, which can better support evolving requirements and scalability goals as compared to Vue.js. This strategic technology shift aligns with industry best practices and positions the firm for long-term growth.

**Comment by tiangolo** [comment, marked as answer]
Let's start with why there was Vue in the first place. :nerd_face: 

The first version of all this, what started this template, long before FastAPI existed, was called `full-stack`, it used Flask and Angular. At the time, Angular had the best support from editors. But it required a lot of complexity and code duplication. So, I moved to Vue.js (Vue 2). It was much simpler, it had great docs, and it had great a very good developer experience, with good editor support. And it had some early attempts at supporting TypeScript. The main components supported it but extensions didn't support it as well. Still, it was better editor support than what React would provide with components based on classes, a bunch of `Component.method.bind(this)` everywhere, and no TypeScript support (React only supported a Facebook-owned type language called Flow or something like that).

But now, years after, React has first-class TypeScript support, and hooks, which solve a lot of the awkwardness of classes and untyped props. And editors have amazing support for JSX (that was surprising to me). So, the developer experience is quite on par with alternatives. I understand Vue 3 has taken a lot of these ideas and in some cases improved them. But still, the difference is not as big as it used to be.

Now, for this project, I want something that people can easily pick up and use, and also something that, if people use it, would have an easier time finding a job. So, even though I'm personally very interested in Vue.js, Solid, and alternatives, for this project React seemed a very good alternative that would tick most of the boxes and strike a good balance, getting most of the benefit possible for everyone.

That's why this now has React. :nerd_face:
